File: April-18/04-04-18/python/01-Help-Patwari.py
MOD = int(1e9)+7
N = 2
K = 2
F = [[1]*(N+1) for i in range(K+1)]
M = [[1]*(N+1) for i in range(K+1)]
F[0][0] = 1
F[0][1] = 1
F[1][0] = 1
F[1][1] = 0

M[0][0] = 1
M[0][1] = 1
M[1][0] = 1
M[1][1] = 0

def fibo(n):
    
    if(n == 1 or n == 0):
        return 1
    power(n)
    return F[0][0]

def power(n):
    if( n == 1 or n == 0):
        return
    power(n//2)
    multiply(F,F)
    if(n % 2 != 0):
        multiply(F,M)

def multiply(F,M):
    a = ( (F[0][0] * M[0][0]) + (F[0][1] * M[1][0]) ) % MOD
    b = ( (F[0][0] * M[0][1]) + (F[0][1] * M[1][1]) ) % MOD
   
    c = ( (F[1][0] * M[0][0]) + (F[1][1] * M[1][0]) ) % MOD
    d = ( (F[1][0] * M[0][1]) + (F[1][1] * M[1][1]) ) % MOD
    
    F[0][0] = a
    F[0][1] = b
    F[1][0] = c
    F[1][1] = d

n = int(input())
print(fibo(n-1))

File: April-18/04-04-18/java/01-Help-Patwari.java
import java.util.*;

class TestClass {
    static int MOD = 1000000007;
    
    static long fibo(long n)
    {
        long F[][] = new long[][]{ {1,1}, {1,0} };
        if(n == 0 || n == 1)
            return 1;
        power(F,n);
        return F[0][0];
    }
    
    static void power(long F[][], long n)
    {
        long M[][] = new long[][]{ {1,1}, {1,0} };
        if(n == 0 || n == 1)
            return;
        power(F, n / 2);
        multiply(F,F);
        if(n % 2 != 0)
            multiply(F,M);
    }
    static void multiply(long F[][], long M[][])
    {
        long a = ( (F[0][0]%MOD) * (M[0][0]%MOD) + (F[0][1]%MOD) * (M[1][0]%MOD) ) % MOD;
        long b = ( (F[0][0]%MOD) * (M[0][1]%MOD) + (F[0][1]%MOD) * (M[1][1]%MOD) ) % MOD;
        
        long c = ( (F[1][0]%MOD) * (M[0][0]%MOD) + (F[1][1]%MOD) * (M[1][0]%MOD) ) % MOD;
        long d = ( (F[1][0]%MOD) * (M[0][1]%MOD) + (F[1][1]%MOD) * (M[1][1]%MOD) ) % MOD;
        
        F[0][0] = a;
        F[0][1] = b;
        F[1][0] = c;
        F[1][1] = d;
    }
    
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        int n = input.nextInt();
        System.out.println( fibo(n-1) );
    }
}
